shadowsocks/integration_test.go

```diff
@@ -0,0 +1,430 @@
+// Copyright 2020 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shadowsocks
+
+import (
+	"bytes"
+	"io"
+	"net"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	logging "github.com/op/go-logging"
+)
+
+func startTCPEchoServer(t testing.TB) *net.TCPListener {
+	listener, err := net.ListenTCP("tcp", &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 0})
+	if err != nil {
+		t.Fatalf("ListenTCP failed: %v", err)
+	}
+	go func() {
+		for {
+			clientConn, err := listener.AcceptTCP()
+			if err != nil {
+				t.Logf("AcceptTCP failed: %v", err)
+				return
+			}
+			go io.Copy(clientConn, clientConn)
```

Comment:
I think you should call clientConn.Close() after the Copy is done.

Reply:
Done